File: ds/list.go
package ds

import (
	"fmt"
)


type ListNode struct {
	Value string
	Next  *ListNode
}

type MyList struct {
	Head *ListNode
}

func NewMyList() *MyList {
	return &MyList{}
}

func (l *MyList) AddHead(val string) {
	l.Head = &ListNode{Value: val, Next: l.Head}
}

func (l *MyList) AddTail(val string) {
	newNode := &ListNode{Value: val}
	if l.Head == nil {
		l.Head = newNode
		return
	}
	cur := l.Head
	for cur.Next != nil {
		cur = cur.Next
	}
	cur.Next = newNode
}

func (l *MyList) AddBefore(target, val string) {
	if l.Head == nil {
		fmt.Println("Список пуст!")
		return
	}
	if l.Head.Value == target {
		l.AddHead(val)
		return
	}
	cur := l.Head
	for cur.Next != nil && cur.Next.Value != target {
		cur = cur.Next
	}
	if cur.Next == nil {
		fmt.Println("Элемент не найден")
		return
	}
	newNode := &ListNode{Value: val, Next: cur.Next}
	cur.Next = newNode
}

func (l *MyList) AddAfter(target, val string) {
	if l.Head == nil {
		fmt.Println("Список пуст")
		return
	}
	cur := l.Head
	for cur != nil && cur.Value != target {
		cur = cur.Next
	}
	if cur == nil {
		fmt.Println("Элемент не найден")
		return
	}
	newNode := &ListNode{Value: val, Next: cur.Next}
	cur.Next = newNode
}

func (l *MyList) DelHead() {
	if l.Head == nil {
		return
	}
	l.Head = l.Head.Next
}

func (l *MyList) DelTail() {
	if l.Head == nil {
		fmt.Println("Список пуст")
		return
	}
	if l.Head.Next == nil {
		l.Head = nil
		return
	}
	cur := l.Head
	for cur.Next.Next != nil {
		cur = cur.Next
	}
	cur.Next = nil
}

func (l *MyList) DelByValue(val string) {
	if l.Head == nil {
		fmt.Println("Список пуст.")
		return
	}
	if l.Head.Value == val {
		l.DelHead()
		fmt.Printf("Элемент %s удален.\n", val)
		return
	}
	cur := l.Head
	for cur.Next != nil && cur.Next.Value != val {
		cur = cur.Next
	}
	if cur.Next == nil {
		fmt.Printf("Элемент %s не найден.\n", val)
		return
	}
	cur.Next = cur.Next.Next
	fmt.Printf("Элемент %s удален.\n", val)
}

func (l *MyList) DelAfterValue(val string) {
	if l.Head == nil || l.Head.Next == nil {
		return
	}
	cur := l.Head
	for cur != nil && cur.Value != val {
		cur = cur.Next
	}
	if cur == nil || cur.Next == nil {
		return
	}
	cur.Next = cur.Next.Next
}

func (l *MyList) DelBeforeValue(val string) {
	if l.Head == nil || l.Head.Next == nil {
		return
	}
	if l.Head.Next.Value == val {
		l.DelHead()
		return
	}
	cur := l.Head
	for cur.Next != nil && cur.Next.Next != nil && cur.Next.Next.Value != val {
		cur = cur.Next
	}
	if cur.Next == nil || cur.Next.Next == nil {
		return
	}
	cur.Next = cur.Next.Next
}

func (l *MyList) FindValue(val string) bool {
	cur := l.Head
	for cur != nil {
		if cur.Value == val {
			return true
		}
		cur = cur.Next
	}
	return false
}

func (l *MyList) ReadForward() {
	if l.Head == nil {
		fmt.Println("Список пуст")
		return
	}
	fmt.Print("Список вперёд: ")
	for cur := l.Head; cur != nil; cur = cur.Next {
		fmt.Printf("\"%s\" ", cur.Value)
	}
	fmt.Println()
}

func (l *MyList) printBackRec(node *ListNode) {
	if node == nil {
		return
	}
	l.printBackRec(node.Next)
	fmt.Printf("\"%s\" ", node.Value)
}

func (l *MyList) ReadBack() {
	if l.Head == nil {
		fmt.Println("Список пуст")
		return
	}
	fmt.Print("Список (назад): ")
	l.printBackRec(l.Head)
	fmt.Println()
}

func (l *MyList) SaveToFile(filename string) error {
    return SaveStringsText(filename, l.toSlice())
}

func (l *MyList) toSlice() []string {
    out := []string{}
    for cur := l.Head; cur != nil; cur = cur.Next {
        out = append(out, cur.Value)
    }
    return out
}


func (l *MyList) LoadFromFile(filename string) error {
    vals, err := LoadStringsText(filename)
    if err != nil { return err }
    l.Head = nil
    for _, v := range vals {
        l.AddTail(v)
    }
    return nil
}

func (l *MyList) SaveToBinaryFile(filename string) error {
    return SaveStringsBinary(filename, l.toSlice())
}

func (l *MyList) LoadFromBinaryFile(filename string) error {
    vals, err := LoadStringsBinary(filename)
    if err != nil { return err }
    l.Head = nil
    for _, v := range vals {
        l.AddTail(v)
    }
    return nil
}

File: list.cpp
#include "list.h"
#include <iostream>
#include <fstream>

using namespace std;

MyList::MyList() : head(nullptr) {}

MyList::~MyList() {
    while (head) delHead();
}

void MyList::addHead(const string& val) {
    head = new Node(val, head);
}

void MyList::addTail(const string& val) {
    if (!head) {
        head = new Node(val, nullptr);
        return;
    }
    Node* curr = head;
    while (curr->next) curr = curr->next;
    curr->next = new Node(val, nullptr);
}

void MyList::delHead() {
    if (!head) {
        cout << "Список пуст. Удаление невозможно." << endl;
        return;
    }
    Node* temp = head;
    head = head->next;
    delete temp;
}

void MyList::addAfter(const string& target, const string& val) {
    Node* curr = head;
    while (curr && curr->value != target) curr = curr->next;
    if (!curr) {
        cout << "Элемент '" << target << "' не найден." << endl;
        return;
    }
    curr->next = new Node(val, curr->next);
}

void MyList::addBefore(const string& target, const string& val) {
    if (!head) {
        cout << "Список пуст. Добавление невозможно." << endl;
        return;
    }
    if (head->value == target) {
        addHead(val);
        return;
    }
    Node* curr = head;
    while (curr->next && curr->next->value != target) curr = curr->next;
    if (!curr->next) {
        cout << "Элемент '" << target << "' не найден." << endl;
        return;
    }
    curr->next = new Node(val, curr->next);
}

void MyList::delTail() {
    if (!head) {
        cout << "Список пуст. Удаление невозможно." << endl;
        return;
    }
    if (!head->next) {
        delete head;
        head = nullptr;
        return;
    }
    Node* curr = head;
    while (curr->next->next) curr = curr->next;
    delete curr->next;
    curr->next = nullptr;
}

void MyList::delByValue(const string& val) {
    if (!head) {
        cout << "Список пуст. Удаление невозможно." << endl;
        return;
    }
    if (head->value == val) {
        delHead();
        return;
    }
    Node* curr = head;
    while (curr->next && curr->next->value != val) curr = curr->next;
    if (!curr->next) {
        cout << "Элемент '" << val << "' не найден." << endl;
        return;
    }
    Node* temp = curr->next;
    curr->next = curr->next->next;
    delete temp;
}

bool MyList::findValue(const string& val) const {
    Node* curr = head;
    while (curr) {
        if (curr->value == val) return true;
        curr = curr->next;
    }
    return false;
}

void MyList::readForward() const {
    if (!head) {
        cout << "Список пуст." << endl;
        return;
    }
    cout << "Список: ";
    Node* curr = head;
    while (curr) {
        cout << curr->value;
        if (curr->next) cout << " -> ";
        curr = curr->next;
    }
    cout << endl;
}

void MyList::printBack(Node* node) const {
    if (!node) return;
    printBack(node->next);
    cout << node->value;
    if (node != head) cout << " <- ";
}

void MyList::readBack() const {
    if (!head) {
        cout << "Список пуст." << endl;
        return;
    }
    cout << "Список (обратный): ";
    printBack(head);
    cout << endl;
}

void MyList::delAfterValue(const string& val) {
    Node* curr = head;
    while (curr && curr->value != val) curr = curr->next;
    if (!curr || !curr->next) {
        cout << "Невозможно удалить элемент после '" << val << "'." << endl;
        return;
    }
    Node* temp = curr->next;
    curr->next = curr->next->next;
    delete temp;
}

void MyList::delBeforeValue(const string& val) {
    if (!head || !head->next || head->value == val) {
        cout << "Невозможно удалить элемент перед '" << val << "'." << endl;
        return;
    }
    if (head->next->value == val) {
        delHead();
        return;
    }
    Node* curr = head;
    while (curr->next->next && curr->next->next->value != val) curr = curr->next;
    if (!curr->next->next) {
        cout << "Элемент '" << val << "' не найден или перед ним нет элемента." << endl;
        return;
    }
    Node* temp = curr->next;
    curr->next = curr->next->next;
    delete temp;
}
